### PR TITLE
Allow the use of custom keys in BPF_HASH_OF_MAPS

### LINKED ISSUE #3496: Add macro HASH_OF_MAPS other types of the key
Hi, `BPF_HASH_OF_MAPS` macro only supports int type key.

I don't know why it doesn't support it, but I wanna use other types of the key. 

I found this [post](https://stackoverflow.com/questions/61214099/functionality-of-the-bpf-hash-of-maps-structure), so maybe we can create `BPF_HASH_OF_MAPS` by using `BPF_TABLE`. But the method isn't clear. And I got some error. So supporting other types of the key is better I think.

The macro definition is [here](https://github.com/iovisor/bcc/blob/2fa54c0bd388898fdda58f30dcfe5a68d6715efc/src/cc/export/helpers.h#L294-L295)
```
#define BPF_HASH_OF_MAPS(_name, _inner_map_name, _max_entries) \
BPF_TABLE("hash_of_maps$" _inner_map_name, int, int, _name, _max_entries)
```

### PR DESCRIPTION
**Pull request by masibw**
close #3496 
Hi team. I wrote a macro that can use custom key for `BPF_HASH_OF_MAPS`.

But in c++,   `get_map_in_map_table` is shared to `BPF_ARRAY_OF_MAPS`.
So, when using `BPF_ARRAY_OF_MAPS`, we have to declare like this.
```
auto t = bpf.get_map_in_map_table<int>("maps_array");
```
Is it okay?

-- solved --
Maybe I could wrote about `BPF_HASH_OF_MAPS` macro. But `.cc` files have errors. I've never write c++ programs, Can anyone help me?
```Bash
In file included from /home/*/bcc/tests/cc/test_map_in_map.cc:21:
In file included from /home/*/bcc/src/cc/api/BPF.h:25:
/home/*/bcc/src/cc/api/BPFTable.h:487:15: error: function 'ebpf::BPFMapInMapTable<custom_key>::update_value' is used but not defined in this translation unit, and cannot be defined in any other translation unit because its type does not have linkage
  StatusTuple update_value(const KeyType& key, const int& inner_map_fd);
              ^
/home/*/bcc/tests/cc/test_map_in_map.cc:182:13: note: used here
    res = t.update_value(hash_key, ex1_fd);
```

This is probably due to the fact that the definition and implementation are in separate files, but if I write it in the header file, the following error occurs. I don't know which one is closer to the solution, so I pushed the one where the definition and implementation are in separate files.

```Bash
In file included from /home/*/bcc/tests/cc/test_map_in_map.cc:21:
In file included from /home/*/bcc/src/cc/api/BPF.h:25:
/home/*/bcc/src/cc/api/BPFTable.h:499:23: error: const_cast from 'char *(*)(const char *, int) throw()' to 'custom_key *' is not allowed
    if (!this->remove(const_cast<KeyType*>(&index)))
```

### CODE REVIEW

**Comment by chenhengqi** [comment]
The implementation of a template class should be in header file.

**Comment by masibw** [comment]
Thanks for replying @chenhengqi, like this code?

but I have this error.

```
In file included from /home/*/bcc/tests/cc/test_map_in_map.cc:21:
In file included from /home/*/bcc/src/cc/api/BPF.h:25:
/home/*/bcc/src/cc/api/BPFTable.h:499:23: error: const_cast from 'char *(*)(const char *, int) throw()' to 'custom_key *' is not allowed
    if (!this->remove(const_cast<KeyType*>(&index)))
```

**Comment by chenhengqi** [comment]
`remove(&key)` not `remove(&index)`

File: src/cc/api/BPF.h [hunk 1]
```diff
@@ -211,8 +211,13 @@ class BPF {
...
                                               bool check_debug_file_crc = true);
-
-  BPFMapInMapTable get_map_in_map_table(const std::string& name);
+  template <class KeyType>
+  BPFMapInMapTable<KeyType> get_map_in_map_table(const std::string& name){
+      TableStorage::iterator it;
+      if (bpf_module_->table_storage().Find(Path({bpf_module_->id(), name}),it))
+        return BPFMapInMapTable<KeyType>(it->second);
+      return BPFMapInMapTable<KeyType>({});
+  }
```
Comment (chenhengqi) [inline src/cc/api/BPF.h:220]: style: 2 space for indention, 1 space after comma.

Comment (masibw) [reply src/cc/api/BPF.h]: Thanks, fixed

**Comment by masibw** [comment]
>remove(&key) not remove(&index)

Oh, I'm sorry. I made a terrible mistake. 😢  
Thanks to you, it's working now.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by masibw** [comment]
Sorry, I made a mistake with the `test_map_in_map.py` file name.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
test_map_in_map.py failed, probably due to python3 compatibility.
```
42: Traceback (most recent call last):
42:   File "/home/fedora/jenkins/workspace/bcc-pr/label/fc28/tests/python/test_map_in_map.py", line 160, in test_hash_table_custom_key
42:     assert(ex1_map[ct.c_int(0)] >= 1)
42: TypeError: '>=' not supported between instances of 'c_int' and 'int'
```

**Comment by yonghong-song** [comment]
[buildbot, ok to test]

**Comment by masibw** [comment]
> test_map_in_map.py failed, probably due to python3 compatibility.

OK, probably fixed.

**Comment by yonghong-song** [comment]
LGTM. Thanks!

**Pull request merged**